### PR TITLE
Fix invalid customization and Customizing the webconsole prompt

### PR DESCRIPTION
**Pull request by cn-kali-team**
# Fix invalid customization
## Steps to reproduce
- You can change the ‘msf’ in the code to any character,like 'KT'

``` ruby
# lib/msf/ui/web/console.rb
# Initialize the console with our pipe
self.console = Msf::Ui::Console::Driver.new(
  'KT',
  '#',
  opts.merge({
    'Framework'   => self.framework,
    'LocalInput'  => self.pipe,
    'LocalOutput' => self.pipe,
    'AllowCommandPassthru' => true,
    'Resource'    => [],
  })
)
```
- start json-rpc service
```
bundle exec thin --rackup msf-json-rpc.ru --address rpc.kali-team.cn --port 8081 --ssl --ssl-key-file /home/kali-team/.msf4/msf-ws-key.pem --ssl-cert-file /home/kali-team/.msf4/msf-ws-cert.pem --ssl-disable-verify --environment development --log /home/kali-team/.msf4/logs/msf-ws.log --pid /home/kali-team/.msf4/msf-ws.pid --tag msf-json-rpc --debug start --threaded
```
- create a console
```
curl --request POST \
  --url https://rpc.kali-team.cn:8081/api/v1/json-rpc \
  --header 'authorization: Bearer 4609b894a5df47bc165f33ae9359ddf4b077fe83803fc390848d45652d4991b7c918c2055a9b3654' \
  --header 'content-type: application/json' \
  --data '{
	"jsonrpc": "2.0",
	"method": "console.create",
	"id": 0
}'
```
- list console
```
curl --request POST \
  --url https://rpc.kali-team.cn:8081/api/v1/json-rpc \
  --header 'authorization: Bearer 4609b894a5df47bc165f33ae9359ddf4b077fe83803fc390848d45652d4991b7c918c2055a9b3654' \
  --header 'content-type: application/json' \
  --data '{
	"jsonrpc": "2.0",
	"method": "console.list",
	"id": 0
}'
```
- response
```
{
  "jsonrpc": "2.0",
  "result": {
    "consoles": [
      {
        "id": "0",
        "prompt": "msf5 > ",
        "busy": false
      }
    ]
  },
  "id": 0
}
```
- The default prompt will be returned no matter what character is modified
- After fixing the bug
```
{
  "jsonrpc": "2.0",
  "result": {
    "consoles": [
      {
        "id": "0",
        "prompt": "web_msf > ",
        "busy": false
      }
    ]
  },
  "id": 0
}
```
# Customizing the webconsole prompt
- create console,The custom prompt is 'KT', prompt_char is '#'
```
curl --request POST \
  --url https://rpc.kali-team.cn:8081/api/v1/json-rpc \
  --header 'authorization: Bearer 4609b894a5df47bc165f33ae9359ddf4b077fe83803fc390848d45652d4991b7c918c2055a9b3654' \
  --header 'content-type: application/json' \
  --data '{
	"jsonrpc": "2.0",
	"method": "console.create",
	"params": [
		{
			"PROMPT": "KT",
			"PROMPTCHAR": "#"
		}
	],
	"id": 0
}'
```
- response
```
{
  "jsonrpc": "2.0",
  "result": {
    "consoles": [
      {
        "id": "0",
        "prompt": "web_msf > ",
        "busy": false
      },
      {
        "id": "1",
        "prompt": "KT # ",
        "busy": false
      }
    ]
  },
  "id": 0
}
```
Thank you for your guidance @timwr

### CODE REVIEW

File: lib/msf/ui/web/console.rb [hunk 1]
```diff
@@ -58,9 +58,9 @@ def initialize(framework, console_id, opts={})
     end
 
     # Initialize the console with our pipe
-    self.console = Msf::Ui::Console::Driver.new(
-      'msf',
-      '>',
+    web_prompt = opts['PROMPT'] || 'web_msf'
```
Comment (smcintyre-r7) [inline lib/msf/ui/web/console.rb:61]: The default should really stay as `'msf'`.

**Review by smcintyre-r7 — CHANGES_REQUESTED** [review]
There seems to be a bug introduced here where once a module is used, the prompt appends to itself each time enter is selected.

```
 : metasploit-framework.pr: 1348911:43:09 metasploit-framework.pr ./msfconsole
                                                  

MMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMM
MMMMMMMMMMM                MMMMMMMMMM
MMMN$                           vMMMM
MMMNl  MMMMM             MMMMM  JMMMM
MMMNl  MMMMMMMN       NMMMMMMM  JMMMM
MMMNl  MMMMMMMMMNmmmNMMMMMMMMM  JMMMM
MMMNI  MMMMMMMMMMMMMMMMMMMMMMM  jMMMM
MMMNI  MMMMMMMMMMMMMMMMMMMMMMM  jMMMM
MMMNI  MMMMM   MMMMMMM   MMMMM  jMMMM
MMMNI  MMMMM   MMMMMMM   MMMMM  jMMMM
MMMNI  MMMNM   MMMMMMM   MMMMM  jMMMM
MMMNI  WMMMM   MMMMMMM   MMMM#  JMMMM
MMMMR  ?MMNM             MMMMM .dMMMM
MMMMNm `?MMM             MMMM` dMMMMM
MMMMMMN  ?MM             MM?  NMMMMMN
MMMMMMMMNe                 JMMMMMNMMM
MMMMMMMMMMNm,            eMMMMMNMMNMM
MMMMNNMNMMMMMNx        MMMMMMNMMNMMNM
MMMMMMMMNMMNMMMMm+..+MMNMMNMNMMNMMNMM
        https://metasploit.com


       =[ metasploit v5.0.90-dev-d2d30282f7               ]
+ -- --=[ 2017 exploits - 1098 auxiliary - 343 post       ]
+ -- --=[ 562 payloads - 45 encoders - 10 nops            ]
+ -- --=[ 7 evasion                                       ]

Metasploit tip: Use the edit command to open the currently active module in your editor

msf5 > use exploit/windows/smb/psexec
msf5 exploit(windows/smb/psexec) > 
msf5 exploit(windows/smb/psexec) exploit(windows/smb/psexec) > 
msf5 exploit(windows/smb/psexec) exploit(windows/smb/psexec) exploit(windows/smb/psexec) > 
msf5 exploit(windows/smb/psexec) exploit(windows/smb/psexec) exploit(windows/smb/psexec) exploit(windows/smb/psexec) > 


```

**Comment by acammack-r7** [comment]
I think the real bug here was that we were adding some ansi codes to the prompt even when it didn't support color, under the assumption (previously mostly true) that anything that had a prompt supported at least _some_ escape codes or could discard them gracefully. https://github.com/rapid7/rex-text/pull/29 Should fix it at the source, without needing to complicate the prompt logic.

Having a way to customize the prompt in the web console is nice.

**Comment by cn-kali-team** [comment]
> I think the real bug here was that we were adding some ansi codes to the prompt even when it didn't support color, under the assumption (previously mostly true) that anything that had a prompt supported at least _some_ escape codes or could discard them gracefully. [rapid7/rex-text#29](https://github.com/rapid7/rex-text/pull/29) Should fix it at the source, without needing to complicate the prompt logic.
> 
> Having a way to customize the prompt in the web console is nice.

Well done, I'll rewrite the code logic when your PR merges

File: lib/msf/ui/web/console.rb (hunk 1 above)
Comment (gwillcox-r7) [reply lib/msf/ui/web/console.rb]: Marking this as resolved as this change was reverted with https://github.com/rapid7/metasploit-framework/pull/13489/commits/4c32d3f34e61f91819ea80aa958c5e9e3e8ee887

**Comment by smcintyre-r7** [comment]
rapid7/rex-text#29 was landed a while ago but it sounds like this is still pending some changes. @cn-kali-team is that right or is this ready for testing again?

**Comment by smcintyre-r7** [comment]
I'm going to close this out with the `attic` label since this doesn't appear to be actively being worked on. If you get back around to it just let us know and we'll happily reopen it! Thanks!

**Pull request closed without merging**